### ISSUE TITLE
mdl-1.x: why there is so few usage of mdlDowngrade_ ?

### ISSUE THREAD

**Issue by mh-cbon** [google/material-design-lite#4921]
Hi,

Can you take a moment to explain the reason that leads to not implement the mdlDowngrade_ on many components ? quickly here i see textfield, tooltip, datatable, button.

I dont quiet get the reasons, also the various issues (#271 #886 #1818) tells it was handled. But the code does not show any of it, [this](https://github.com/google/material-design-lite/blob/mdl-1.x/src/slider/slider.js).

What am i missing here ?

**Comment by Garbee** [comment]
We did away with downgrading bits as so long as the references are cleaned up the engine will clear out any bound even listeners when the nodes are removed.

The reason things lagged around earlier, was the `componentHandler` kept an internal reference to things that wasn't caught with initial testing. Which led us to thinking the downgrade handlers were needed when in fact they weren't. We fixed that bug by removing the internal reference on downgrade, so nodes can be downgraded then removed from the DOM and the memory leaks won't occur like they were before.

**Comment by mh-cbon** [comment]
Thanks for the feedback.

Did you kept around the methodology to detect memory leaks ?

**Comment by mh-cbon** [comment]
Maybe the doc here should be fixed https://github.com/google/material-design-lite/wiki/Making-your-first-JS-component :x

**Comment by Garbee** [comment]
Wiki is fixed on that front.

Yes, we *completely removed* the downgrade check and execution as it is no longer needed in core. It was a leaky abstraction that served no purpose once we found the actual source of the problem. It was a minor BC break for any existing components that required that, but it was best removed to keep any future debugging of the manager as straightforward as could be.

**Comment by mh-cbon** [comment]
tbh, not sure i totally understand the whole story behind this decision. But, i won t argue about it, i made my own sauce (https://github.com/mh-cbon/material-design-lite/commit/e206e7b317ec7c1dc0512038c4c8182f7db2bde3#diff-8118a6483928b8c7f76329f0d99096b7) because i do need to downgrade / upgrade properly the components (see the custom-dup components for more understanding).

Although i only regret there is not a clear method to check those matters.

thanks again.

**Comment by Garbee** [comment]
The idea is you shouldn't really need to downgrade things just to re-upgrade them. You downgrade them in order to remove them from the DOM. Once we realized the memory leak detected which caused the addition of the downgrade check was actually a bug internally, we addressed the bug and decided to remove the check to keep what we are maintaining minimal. The component handler is non-trivial and for every extra thing it does when someone brings forward a bug it creates that much more complexity to go over testing.

Basically, there is no need to have a downgrade method in most applications in the space we are targeting (more static web-pages and not fully-dynamic web apps.) However, even with most modern setups they destroy and rebuild the full DOM of a component on changes. So, the downgrading as it is still works in those cases.
